main.cpp

```diff
@@ -102,18 +103,38 @@ void uthread_create(void *(start_routine)(void *), void* arg){
     tcb->arg = arg;
 
     // Allocate the stack.
-    auto* stack = (char*)malloc(STACK_SIZE);
+    auto* stack = (char *)malloc(STACK_SIZE);
 
     // sp starts out at the top of the stack, pc at the wrapper function.
     auto sp = (address_t)stack + STACK_SIZE - 10 * sizeof(void*);
     auto pc = (address_t)thread_wrapper;
 
+    // Place the argument and the function pointer on the stack
+
+    address_t *stack_p = (address_t *)sp;
+
+    *stack_p = (address_t)arg;
+    sp--;
+    stack_p = (address_t *)sp;
+
+    *stack_p = (address_t)start_routine;
+    sp--;
+    stack_p = (address_t *)sp;
+
+    // Place the wrapper function on the stack so when the scheduler switches
+    // to this thread, the wrapper function is popped off the stack
+    *stack_p = (address_t)thread_wrapper;
+    sp--;
+
```

Comment:
This should work now. But I'm not even sure if we even need this because it doesn't seem to have much effect on the program when it runs.

Reply:
It seems like siglongjump does all of this work for us